nomad/csi_endpoint.go

```diff
@@ -1118,6 +1182,112 @@ func (v *CSIVolume) createVolume(vol *structs.CSIVolume, plugin *structs.CSIPlug
 	return nil
 }
 
+// expandVolume validates the requested capacity values and issues
+// ControllerExpandVolume (and NodeExpandVolume, if needed) to the CSI plugin,
+// via Nomad client RPC.
+//
+// Note that capacity can only be increased; reduction in size is not possible,
+// and if the volume is already at the desired capacity, no action is taken.
+// vol Capacity-related values are mutated if successful, so callers should
+// pass in a copy, then commit changes to raft.
+func (v *CSIVolume) expandVolume(vol *structs.CSIVolume, plugin *structs.CSIPlugin, capacity *csi.CapacityRange) error {
+	if vol == nil || plugin == nil || capacity == nil {
+		return errors.New("unexpected nil value")
+	}
+
+	newMax := capacity.LimitBytes
+	newMin := capacity.RequiredBytes
+
+	// closure for verbose logging, including changes to vol fields
+	logger := func() hclog.Logger { // TODO: keep or nah?
```

Comment:
I like the additional detail, especially since CSI can be tricky in the first place. Do we need the closure though? Can we just assign logger via the `.Named().With()` call and then use this newly created sub-logger?

Reply:
I made it a closure because the fields on `vol` change over the course of `expandVolume()`, and being a closure allows those changes to be reflected in the log lines.  It felt useful to me, but it's not a common pattern from what I've seen, so I'm on the fence. :shrug:

Reply:
Changing the values feels like it's going to be confusing, as the logged values won't be what the user sent as input.

Reply:
the user's input are here as "new_min" and "new_max" and those do not change along the way.  the `vol` values ("current_min", "current_max", and "capacity") in the happy case would change to align with the "new_*"s

Reply:
Right, but if we're returning an error then we'll have removed the information about what we did.

Reply:
not sure I follow you here.

Reply:
I'm guess I'm mostly looking at this from the perspective of error logs, because roughly nobody runs with debug logs in production. And I'm saying that we can populate all the values at the start of the RPC handler, so that when the user (well, administrator) sees the error log they can see "this is what I asked for" and "this was the error" and not some intermediate state where the values are changed from what they knew them to be. The log message itself has the context about whether something has changed or not.

_But_ it turns out you never log errors here so that's totally moot! I'm not sure there's a lot of value in repeating these values across 6 different log messages though... we're sort of using these as a status check and the values won't change until the very last line. So why not put them at the start and stop only?

Reply:
The values actually change twice along the way
* `vol.RequestedCapacity(Min|Max)` are set right after they are validated as logically coherent
* `vol.Capacity` changes after the controller has done the actual expansion

But as for "status check"s, yeah these are probably more like `Trace`s, really, but I figured this operation will probably happen pretty rarely, and might be nice to be able to debug without all the other trace noise.

All that said, it sounds like the feeling is that this is overkill for not much gain, eh?  I can boil it down to just the requested values and structure it more straightforwardly.